app/lib/frontend/dom/dom.dart

```diff
@@ -0,0 +1,160 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:convert';
+
+final _attributeEscape = HtmlEscape(HtmlEscapeMode.attribute);
+
+/// The DOM context to use while constructing nodes.
+///
+/// Override this in browser.
+DomContext dom = StringDomContext();
+
+/// Opaque entity for DOM nodes.
+abstract class Node {}
+
+/// Factory class to create DOM [Node]s.
+abstract class DomContext {
+  /// Creates a DOM fragment from the list of [children] nodes.
+  Node fragment(List<Node> children);
+
+  /// Creates a DOM Element.
+  Node element(
+    String tag, {
+    String id,
+    List<String> classes,
+    Map<String, String> attributes,
+    List<Node> children,
+  });
```

Comment:
Consider using `Iterable<T>` in interfaces instead of `List<T>`.
```suggestion
  Node fragment(Iterable<Node> children);

  /// Creates a DOM Element.
  Node element(
    String tag, {
    String id,
    Iterable<String> classes,
    Map<String, String> attributes,
    Iterable<Node> children,
  });
```